### ISSUE TITLE
What is delay_allreduce and when to use it?

### ISSUE THREAD

**Issue by BramVanroy** [NVIDIA/apex#539]
I was looking through the documentation

https://github.com/NVIDIA/apex/blob/088985936518be7e25795a30d8ab33affa9db6ed/apex/parallel/distributed.py#L151

And some examples

https://github.com/NVIDIA/apex/blob/088985936518be7e25795a30d8ab33affa9db6ed/examples/imagenet/main_amp.py#L163-L168

but I am still unclear when delay_allreduce should be used. It doesn't seem to be required, but what are its benefits? When **must** one use it?

**Comment by mcarilli** [comment]
With overlapped comms (delay_allreduce=False, the default), Apex DDP determines which gradients to expect based on which weights have `requires_grad=True`.  If some parameters with `requires_grad=True` are not actually used, the allreduces in the backward pass will hang, waiting for gradients that never come.

`delay_allreduce=True` is intended to support such cases, by waiting until the end of backward then allreducing all gradients that happened to be created.

The current `torch.nn.parallel.DistributedDataParallel` supports such cases in a more clever way, by tracing the graph built by the forward pass and determining which parameters actually expect gradients.

The bottom line is, I recommend using `torch.nn.parallel.DistributedDataParallel` these days.  It's efficient, robust, it overlaps comms, and most importantly it's packaged directly with upstream so there's no danger of breaking changes.  Apex DDP exists mainly to support internal use cases that rely on it (+offers some really marginal gains like the ability to put multiple allreduces in flight at once).  Torch and Apex DDP are drop-in replacements for each other, aside from some [additional constructor arguments](https://github.com/NVIDIA/apex/blob/master/examples/simple/distributed/distributed_data_parallel.py#L49-L52) for Torch DDP.

**Comment by BramVanroy** [comment]
Ah, that makes sense. Thank you for the quick response! I did some testing of fine-tuning distilbert on 4x Tesla V100 and comparing torch DDP vs apex DDP and found minor differences in favor of torch DDP. Leaving this here as a reference.

|                   | run-1 | run-2 | run-3 | run-4 |
|-------------------|-------|-------|-------|-------|
| distributed       | torch | torch | apex  | apex  |
| fp16              | no    | yes   | no    | yes   |
| **training it/s** | 28.08 | 51.27 | 27.80 | 50.31 |

**Comment by nbcsm** [comment]
@mcarilli 
So ```delay_allreduce=True``` is just a workaround/hack to get rid of a potential hang issue. It does not have any performance benefit, instead it may even slow down the training due to no-overlap between compute/communication.

is my understanding right? tks.

**Comment by viperit** [comment]
> @mcarilli
> So `delay_allreduce=True` is just a workaround/hack to get rid of a potential hang issue. It does not have any performance benefit, instead it may even slow down the training due to no-overlap between compute/communication.
> 
> is my understanding right? tks.

yes, you can find 

> [here](https://github.com/pytorch/pytorch/pull/24800)
